Cython/Utility/TypeConversion.c

```diff
@@ -290,15 +290,22 @@ static CYTHON_INLINE const char* __Pyx_PyObject_AsStringAndSize(PyObject* o, Py_
         return PyByteArray_AsString(o);
 #endif
 
-    } else
-    {
+    } else if (PyBytes_Check(o)) {
         char* result;
         int r = PyBytes_AsStringAndSize(o, &result, length);
         if (unlikely(r < 0)) {
             return NULL;
         } else {
             return result;
         }
+    } else {
+        char* result;
+        int r = PyArg_Parse(o, "y#", &result, &length);
```

Comment:
Used [`PyArg_Parse`]( https://docs.python.org/3/c-api/arg.html#c.PyArg_Parse ) as it looks like that can take a single value

Also from [the "strings and buffers" section]( https://docs.python.org/3/c-api/arg.html#strings-and-buffers ) used `y#` as it looks like it is limited to `bytes`-like objects where a pointer and size can be returned

Reply:
Should add was able to test this out locally and it worked ok

Reply:
I was generally unimpressed by the performance of the `PyArg_*()` functions and tend to avoid them if possible. Not sure if it makes a difference here, given that the most important cases are already handled up-front and we end up asking the owner to fill our buffer struct etc. These things usually tend to harm smaller data buffers more than larger ones.

If it turns out to be a visible factor, we could try to inline the relevant code for `y#` specifically. Cython has helper macros to look up slots in both the normal and Limited C-API.

Reply:
The only thing there is hat you can't use `PyType_GetSlot` to "backdate" support in the Limited API so I think `PyArg_Parse` may be the only option there. Agree with backdating it elsewhere.

Reply:
Tried the following approaches

```cython
from cpython.getargs cimport PyArg_Parse
from cpython.buffer cimport (
    PyBUF_SIMPLE, Py_buffer,
    PyObject_GetBuffer, PyBuffer_Release,
)


cpdef noop_to_c_string(b):
    pass


cpdef pybuf_to_c_string(b):
    cdef Py_buffer pb
    cdef const char* s
    cdef Py_ssize_t n
    try:
        PyObject_GetBuffer(b, &pb, PyBUF_SIMPLE)
        s = <const char*>pb.buf
        n = pb.len
    finally:
        PyBuffer_Release(&pb)


cpdef pyarg_to_c_string(b):
    cdef const char* s
    cdef Py_ssize_t n
    PyArg_Parse(b, "y#", &s, &n)
```

Then used the string `b"abc"` to get the following results

```python
%timeit noop_to_c_string(b)
15.3 ns ± 0.0764 ns per loop (mean ± std. dev. of 7 runs, 100,000,000 loops each)

%timeit pybuf_to_c_string(b)
18.7 ns ± 0.259 ns per loop (mean ± std. dev. of 7 runs, 100,000,000 loops each)

%timeit pyarg_to_c_string(b)
31.3 ns ± 0.31 ns per loop (mean ± std. dev. of 7 runs, 10,000,000 loops each)
```

So agree it is slower to use `PyArg_Parse` and is probably best saved as a fallback for the limited API on Python pre-3.11

As `PyObject_GetBuffer`, is already calling the slot functions itself and a Python could be compiled with build time optimizations, likely that is about as fast as gets

Can add a compile time branch in this `else` to capture these two cases

Reply:
Actually another option would be to use [the old-style buffer protocol]( https://docs.python.org/3.12/c-api/objbuffer.html#old-buffer-protocol ). This is part of the stable API and was around until Python 3.12

If we did this we could call [`PyObject_AsCharBuffer`]( https://docs.python.org/3.12/c-api/objbuffer.html#c.PyObject_AsCharBuffer ) like so

```suggestion
        int r = PyObject_AsCharBuffer(o, &result, &length);
```

Reply:
> AFAICT it doesn't check whether releasing the buffer is required. It just always releases the buffer

But that doesn't make it any better than what the current master branch does, right?

Reply:
For now have added `Py_buffer` approach for most cases (non-Limited API on any Python version or Python 3.11+ w/Limited API). So that ensures the bulk of cases get the more performant approach. Also added a check that the slot `bf_releasebuffer` is `NULL` and raised an error if not

For all other cases (so Limited API on 3.8, 3.9 or 3.10), we fallback to `PyArg_Parse`. It is slower, but it still provides the new behavior in those cases. Should add [the `PyArg_Parse` case does check that the buffer does not need to be released]( https://github.com/python/cpython/blob/8a00c9a4d2ce9d373b13f8f0a2265a65f4523293/Python/getargs.c#L1245-L1250 )